File: other/permutations/permutations.rb
def get_permutations(string)
  return [''] if string.length == 0

  first = string[0]
  remainder = string.slice(1..-1)
  words = get_permutations(remainder)
  permutations = []
  words.each do |word|
    (word.length + 1).times do |i|
      permutations.push(insert_char_at(first, word, i))
    end
  end
  permutations
end

def insert_char_at(char, word, index)
  word.dup.insert(index, char)
end


p get_permutations('a') == ['a']
p get_permutations('ab') == ['ab','ba']
p get_permutations('abc').sort == ['abc','bac','bca','acb','cab','cba'].sort

File: other/permutations/permutations.js
var getPermutations = function(string) {
  if (string.length === 0) {
    return [''];
  }
  var first = string[0];
  var remainder = string.substring(1);
  var words = getPermutations(remainder);
  var permutations = [];
  for (var i in words) {
    for (var j = 0; j <= i.length; j++ )
      permutations.push(insertChar(first, words[i], j));
  }
  return permutations;
}

var insertChar = function(character, word, index) {
  var head = word.substring(0, index);
  var tail = word.substring(index);
  return head + character + tail;
}

console.log("Permutations of 'abc' are:");
console.log(getPermutations('abc'));